hail/python/test/hail/linalg/test_linalg.py

```diff
@@ -457,6 +457,36 @@ def test_matrix_ops(self):
                                                          [14.0, 16.0, 18.0]]))
         self._assert_eq(square.sum(axis=0).T + square.sum(axis=1), np.array([[18.0], [30.0], [42.0]]))
 
+    def test_tree_matmul(self):
+        nm = np.array([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]])
+        m = BlockMatrix.from_numpy(nm, block_size=2)
+        nrow = np.array([[7.0, 8.0, 9.0]])
+        row = BlockMatrix.from_numpy(nrow, block_size=2)
+
+        self._assert_eq(m.tree_matmul(m.T, 2, new_temp_file()), nm @ nm.T)
+        self._assert_eq(m.tree_matmul(nm.T, 2, new_temp_file()), nm @ nm.T)
+        self._assert_eq(row.tree_matmul(row.T, 2, new_temp_file()), nrow @ nrow.T)
+        self._assert_eq(row.tree_matmul(nrow.T, 2, new_temp_file()), nrow @ nrow.T)
+
+        self._assert_eq(m.T.tree_matmul(m, 2, new_temp_file()), nm.T @ nm)
+        self._assert_eq(m.T.tree_matmul(nm, 2, new_temp_file()), nm.T @ nm)
+        self._assert_eq(row.T.tree_matmul(row, 2, new_temp_file()), nrow.T @ nrow)
+        self._assert_eq(row.T.tree_matmul(nrow, 2, new_temp_file()), nrow.T @ nrow)
```

Comment:
call me paranoid, but it would be nice to have at least one example of x @ y where x and y are both block matrices and neither is the transpose of the other.

Reply:
Heh, I just copied the current matmul tests plus that looping of different block sizes, but you're right, these are only transposes. I'll add another test.